src/Drips.sol

```diff
@@ -358,14 +266,132 @@ library Drips {
         );
     }
 
+    /// @notice Calculates the end time of drips without duration.
+    /// @param balance The balance when drips have started
+    /// @param dripsStart The timestamp when drips have started.
+    /// @param receivers The list of drips receivers.
+    /// Must be sorted, deduplicated and without 0 amtPerSecs.
+    /// @return defaultEndTime The end time of drips without duration.
+    function _defaultEnd(
+        uint128 balance,
+        uint64 dripsStart,
+        DripsReceiver[] memory receivers
+    ) private pure returns (uint64 defaultEndTime) {
+        require(receivers.length <= MAX_DRIPS_RECEIVERS, "Too many drips receivers");
+        DefaultEndReceiver[] memory defaults = new DefaultEndReceiver[](receivers.length);
+        uint256 length = 0;
+
+        for (uint256 i = 0; i < receivers.length; i++) {
+            DripsReceiver memory receiver = receivers[i];
+            require(receiver.amtPerSec != 0, "Drips receiver amtPerSec is zero");
+            require(i == 0 || _isOrdered(receivers[i - 1], receiver), "Receivers not sorted");
+            uint64 start = receiver.start;
+            if (start < dripsStart) start = dripsStart;
+            if (receiver.duration == 0) {
+                length = _addDefaultEndReceiver(defaults, length, start, receiver.amtPerSec);
+            } else {
+                uint64 end = _capTimestamp(uint72(start) + receiver.duration);
+                uint192 spent = (end - start) * receiver.amtPerSec;
+                require(balance >= spent, "Insufficient balance");
+                balance -= uint128(spent);
+            }
+        }
+        return _receiversDefaultEnd(defaults, length, balance);
+    }
+
+    /// @notice The internal representation of a receiver without a duration
+    struct DefaultEndReceiver {
+        uint64 start;
+        uint128 amtPerSec;
+    }
+
+    /// @notice Adds a `DefaultEndReceiver` to the list of receivers while keeping it sorted.
+    /// @param receivers The list of receivers, must be sorted by start
+    /// @param length The length of the list
+    /// @param start The start time of the added receiver
+    /// @param amtPerSec The amtPerSec of the added receiver
+    /// @return newLength The new length of the list
+    function _addDefaultEndReceiver(
+        DefaultEndReceiver[] memory receivers,
+        uint256 length,
+        uint64 start,
+        uint128 amtPerSec
+    ) private pure returns (uint256 newLength) {
+        for (uint256 i = 0; i < length; i++) {
+            DefaultEndReceiver memory receiver = receivers[i];
+            if (receiver.start == start) {
```

Comment:
I think we should call the struct `DefaultEndTimes` because they are not really receivers because we missing the `address`

Reply:
Well, `DefaultEndTime` doesn't contain the end time, so it's not great either. How about just `DefaultEnd`?

Reply:
Fixed